File: problems/2026-06-25/job-idempotent-payment/solution.py
import uuid
from typing import Literal


class PaymentProcessor:
    """
    Idempotent payment processor.

    process(idempotency_key, amount) must be safe to retry:
    - First call processes and stores the result.
    - Repeat calls with the same key return the original result without recharging.
    - Two simultaneous calls with the same key: only one should process.
    """

    def __init__(self):
        # TODO: set up internal store (dict is fine — stands in for Redis)
        pass

    def process(self, idempotency_key: str, amount: float) -> dict:
        """
        Returns:
          { "status": "ok",       "confirmation_id": str }  — new payment
          { "status": "duplicate","confirmation_id": str }  — repeated key (same result)
          { "status": "conflict" }                          — key in-flight (race condition)
        TODO:
        1. Check store for idempotency_key
           - If "completed" → return { status: "duplicate", confirmation_id: ... }
           - If "in-flight"  → return { status: "conflict" }
        2. Mark key as "in-flight"
        3. Fake the charge: confirmation_id = str(uuid.uuid4())
        4. Mark key as "completed" with confirmation_id
        5. Return { status: "ok", confirmation_id: ... }
        """
        pass


# ─── Test Runner ──────────────────────────────────────────────────────────────
def run_tests(cases):
    passed = 0
    for i, (label, result, expected) in enumerate(cases, 1):
        if result == expected:
            passed += 1
            print(f"✓  {i}. {label}")
        else:
            print(f"✗  {i}. {label}")
            print(f"   Expected: {expected}")
            print(f"   Got:      {result}")
    print("\n" + "─" * 40)
    print(f"  {passed}/{len(cases)} test cases passed")


if __name__ == "__main__":
    p = PaymentProcessor()

    first = p.process("key-abc", 50.0)
    repeat = p.process("key-abc", 50.0)
    fresh = p.process("key-xyz", 20.0)

    run_tests([
        ("first call returns ok",
            first.get("status") if first else None,
            "ok"),
        ("repeat call returns duplicate",
            repeat.get("status") if repeat else None,
            "duplicate"),
        ("repeat returns same confirmation_id",
            repeat.get("confirmation_id") if repeat else None,
            first.get("confirmation_id") if first else None),
        ("different key is independent",
            fresh.get("status") if fresh else None,
            "ok"),
        ("different key has its own confirmation_id",
            fresh.get("confirmation_id") != first.get("confirmation_id") if fresh and first else False,
            True),
    ])

File: problems/2026-06-25/job-idempotent-payment/solution.ts
import { randomUUID } from "crypto";

type StoreEntry =
  | { status: "in-flight" }
  | { status: "completed"; confirmationId: string };

type ProcessResult =
  | { status: "ok";        confirmationId: string }
  | { status: "duplicate"; confirmationId: string }
  | { status: "conflict" };

class PaymentProcessor {
  /**
   * Idempotent payment processor.
   *
   * process(idempotencyKey, amount) must be safe to retry:
   * - First call processes and stores the result.
   * - Repeat calls with the same key return the original result without recharging.
   * - Two simultaneous calls with the same key: only one should process.
   */

  // TODO: add a store (Map is fine — stands in for Redis)

  process(idempotencyKey: string, amount: number): ProcessResult {
    // TODO:
    // 1. Check store for idempotencyKey
    //    - If "completed" → return { status: "duplicate", confirmationId }
    //    - If "in-flight"  → return { status: "conflict" }
    // 2. Mark key as "in-flight"
    // 3. Fake the charge: confirmationId = randomUUID()
    // 4. Mark key as "completed" with confirmationId
    // 5. Return { status: "ok", confirmationId }
    return { status: "conflict" };
  }
}


// ─── Test Runner ──────────────────────────────────────────────────────────────
function runTests(cases: Array<[string, unknown, unknown]>): void {
  let passed = 0;
  for (let i = 0; i < cases.length; i++) {
    const [label, result, expected] = cases[i];
    const ok = JSON.stringify(result) === JSON.stringify(expected);
    if (ok) {
      passed++;
      console.log(`✓  ${i + 1}. ${label}`);
    } else {
      console.log(`✗  ${i + 1}. ${label}`);
      console.log(`   Expected: ${JSON.stringify(expected)}`);
      console.log(`   Got:      ${JSON.stringify(result)}`);
    }
  }
  console.log("\n" + "─".repeat(40));
  console.log(`  ${passed}/${cases.length} test cases passed`);
}

const p = new PaymentProcessor();

const first  = p.process("key-abc", 50);
const repeat = p.process("key-abc", 50);
const fresh  = p.process("key-xyz", 20);

runTests([
  ["first call returns ok",
    (first  as any).status, "ok"],
  ["repeat call returns duplicate",
    (repeat as any).status, "duplicate"],
  ["repeat returns same confirmationId",
    (repeat as any).confirmationId, (first as any).confirmationId],
  ["different key is independent",
    (fresh  as any).status, "ok"],
  ["different key has its own confirmationId",
    (fresh as any).confirmationId !== (first as any).confirmationId, true],
]);
